### PR TITLE
[IMP] sale: ask confirmation for SO cancellation

### PR DESCRIPTION
**Pull request by Demesmaeker**
Before this commit, when cancelling a SO, the wizard opened only when
there were draft invoices or delivered pickings in the order.

In order to complete the notification flow of the SO and avoid
unexpected  cancellation of orders, the user must now confirm
cancellation and gets to possibility to send an email notification.

task-2660895

See also:
- https://github.com/odoo/enterprise/pull/26866

### CODE REVIEW

File: addons/sale/models/sale_order.py [hunk 1]
```diff
@@ -914,7 +930,11 @@ def _action_cancel(self):
 
     def _show_cancel_wizard(self):
         for order in self:
-            if order.invoice_ids.filtered(lambda inv: inv.state == 'draft') and not order._context.get('disable_cancel_warning'):
+            if (
+                order.state != 'draft' and not order._context.get('disable_cancel_warning')
+                or order.invoice_ids.filtered(lambda inv: inv.state == 'draft')
+                and not order._context.get('disable_cancel_warning')
+            ):
```
Comment (chevalierv) [inline addons/sale/models/sale_order.py:937]:
```suggestion
               (order.state != 'draft' or order.invoice_ids.filtered(lambda inv: inv.state == 'draft'))
                and not order._context.get('disable_cancel_warning')
            ):
```

File: addons/sale_stock/wizard/sale_order_cancel_views.xml [hunk 2]
```diff
@@ -5,12 +5,12 @@
...
-            <field name="display_invoice_alert" position="after">
+            <span id="display_invoice_alert" position="after">
                 <field name="display_delivery_alert" invisible="1"/>
-                <div attrs="{'invisible': [('display_delivery_alert', '=', False)]}">
+                <span attrs="{'invisible': [('display_delivery_alert', '=', False)]}">
                     Some products have already been delivered. Returns can be created from the Delivery Orders.
-                </div>
-            </field>
+                </span>
+            </span>
```
Comment (chevalierv) [inline addons/sale_stock/wizard/sale_order_cancel_views.xml:13]: Why are those changes needed?

**Review by chevalierv — COMMENTED** [review]
Looks good to me :+1: Good work!

File: addons/sale_stock/wizard/sale_order_cancel_views.xml (hunk 2 above)
Comment (Demesmaeker) [reply addons/sale_stock/wizard/sale_order_cancel_views.xml:13]: To get all 3 messages (in the worst-case scenario) in a single block, instead of 3 different div, it's visually better.

File: addons/sale/models/sale_order.py (hunk 1 above)
Comment (Demesmaeker) [reply addons/sale/models/sale_order.py:941]: Ha, my bad, this isn't even necessary, we have the cancel wizard as long as the quotation isn't in draft (and draft quotation don't have invoice) so the check on the invoice is useless anyway

File: addons/sale/wizard/sale_order_cancel.py [hunk 3]
```diff
@@ -1,20 +1,87 @@
...
 
 
 class SaleOrderCancel(models.TransientModel):
     _name = 'sale.order.cancel'
-    _description = "Sales Order Cancel"
+    _inherit = 'mail.composer.mixin'
+    _description = 'Sales Order Cancel'
 
+    @api.model
+    def _get_default_from(self):
```
Comment (Feyensv) [inline addons/sale/wizard/sale_order_cancel.py:14]:
```suggestion
    def _default_email_from(self):
```

https://www.odoo.com/documentation/14.0/developer/misc/other/guidelines.html#symbols-and-conventions
-> Default method : the default method pattern is `_default_<field_name>`

File: addons/sale/wizard/sale_order_cancel.py [hunk 4]
```diff
@@ -1,20 +1,87 @@
...
+    def action_send_mail_and_cancel(self):
+        for order in self:
+            order.order_id.message_post(
+                subject=order.subject,
+                body=order.body,
+                message_type='comment',
+                email_from=self.author_id.email,
+                email_layout_xmlid='mail.mail_notification_light',
+                partner_ids=self.recipient_ids.ids,
+            )
```
Comment (Feyensv) [inline addons/sale/wizard/sale_order_cancel.py:83]: No need to support multi for the action ;)
--> self.ensure_one() et hop :D

**Review by Feyensv — COMMENTED** [review]
First tech pass :duck:

I noticed you don't modify other modules code (except test/views), but you have to know that some flows trigger the cancellation automatically (e.g. `event_booth_sale -> _cancel_pending_registrations`, probably some flows of ebay/amazon?, ...).  You should probably see with DALA what we should do for those cases.  We do not want to open the wizard in those cases (obviously) but maybe we still want to send an email in some cases :thinking:.

File: addons/sale/data/mail_template_data.xml [hunk 5]
```diff
@@ -231,5 +231,35 @@
...
+        <record id="sale.mail_template_sale_cancellation" model="mail.template">
+            <field name="name">Sales Order: Cancellation email</field>
+            <field name="model_id" ref="sale.model_sale_order"/>
+            <field name="subject">{{ object.company_id.name }} {{ object.state in ('draft', 'sent') and 'Quotation Cancelled' or 'Order Cancelled' }} (Ref {{ object.name or 'n/a' }})</field>
+            <field name="email_from">{{ (object.user_id.email_formatted or user.email_formatted) }}</field>
+            <field name="partner_to">{{ object.partner_id.id }}</field>
+            <field name="body_html" type="html">
+<div style="margin: 0px; padding: 0px;">
+    <p style="margin: 0px; padding: 0px; font-size: 13px;">
+        <t t-set="doc_name" t-value="'quotation' if object.state in ('draft', 'sent') else 'order'"/>
```
[the remark this answers is not in the document]
Comment (Demesmaeker) [reply addons/sale/data/mail_template_data.xml]: Corrected using `object.type_name`

[the remark this answers is not in the document]
Comment (Feyensv) [reply addons/sale/data/mail_template_data.xml]: Oh, didn't see this comment was still there (and hidden) from my old pending review :man_facepalming:

File: addons/sale/data/mail_template_data.xml [hunk 6]
```diff
@@ -231,5 +231,35 @@
...
+            <field name="body_html" type="html">
+<div style="margin: 0px; padding: 0px;">
+    <p style="margin: 0px; padding: 0px; font-size: 13px;">
+        <t t-set="doc_name" t-value="object.type_name"/>
+        Dear <t t-out="object.partner_id.name or ''">user</t>,
+        <br/><br/>
+        Please be advised that your
+        <t t-out="doc_name or ''">quotation</t> <strong t-out="object.name or ''"></strong>
+        <t t-if="object.origin">
+            (with reference: <t t-out="object.origin or ''">S00052</t> )
```
[the remark this answers is not in the document]
Comment (Demesmaeker) [reply addons/sale/data/mail_template_data.xml:250]: This is exactly the logic used in the sale template when sending an mail, I agree the string `with reference` is a bit misleading, but should we also change the logic on the other template then?

[the remark this answers is not in the document]
Comment (Feyensv) [reply addons/sale/data/mail_template_data.xml:250]: Let's ask @dalaOdoo if this part is really needed (& clear) for our use case to see if it's worth keeping :)

[the remark this answers is not in the document]
Comment (dalaOdoo) [reply addons/sale/data/mail_template_data.xml:250]: I don't see reason to change from other templates. If there is a reference we should use it

File: addons/sale/wizard/sale_order_cancel_views.xml [hunk 7]
```diff
@@ -1,18 +1,51 @@
...
+                        <span>Are you sure you want to cancel this order? <br/></span>
+                        <span id="display_invoice_alert"
+                              attrs="{'invisible': [('display_invoice_alert', '=', False)]}">
+                            Draft invoices for this order will be cancelled. <br/>
+                        </span>
+                    </div>
+                    <group col="2">
+                        <field name="recipient_ids"
+                               widget="many2many_tags_email"
+                               attrs="{'readonly': False}"
```
Comment (Feyensv) [inline addons/sale/wizard/sale_order_cancel_views.xml:25]: -> readonly should be specified on the field instead

File: addons/sale/wizard/sale_order_cancel_views.xml [hunk 8]
```diff
@@ -1,18 +1,51 @@
 <?xml version="1.0" encoding="utf-8"?>
 <odoo>
     <record id="sale_order_cancel_view_form" model="ir.ui.view">
-        <field name="name">sale.order.cancel.form</field>
+        <field name="name">sale.sale.order.cancel.form</field>
```
Comment (Feyensv) [inline addons/sale/wizard/sale_order_cancel_views.xml:4]: No need to modify this line, the current logic model_name + view_type is clear enough, no need to add the module ;)

File: addons/sale/wizard/sale_order_cancel.py [hunk 9]
```diff
@@ -1,20 +1,87 @@
...
+                    post_process=True,
+                )[wizard.order_id.id]
+
+    @api.depends('order_id')
+    def _compute_body(self):
+        for wizard in self:
+            if wizard.template_id:
+                wizard.body = self.sudo()._render_template(
+                    wizard.template_id.body_html,
+                    'sale.order', [wizard.order_id.id],
```
Comment (Feyensv) [inline addons/sale/wizard/sale_order_cancel.py:69]:
```suggestion
                    'sale.order',
                    [wizard.order_id.id],
```

File: addons/sale/wizard/sale_order_cancel.py [hunk 10]
```diff
@@ -1,20 +1,87 @@
...
+    )
+
+    # recipients
+    recipient_ids = fields.Many2many(
+        'res.partner', string='Recipients', compute='_compute_recipient_ids'
+    )
     order_id = fields.Many2one('sale.order', string='Sale Order', required=True, ondelete='cascade')
-    display_invoice_alert = fields.Boolean('Invoice Alert', compute='_compute_display_invoice_alert')
+    display_invoice_alert = fields.Boolean(
+        'Invoice Alert', compute='_compute_display_invoice_alert'
```
Comment (Feyensv) [inline addons/sale/wizard/sale_order_cancel.py:35]:
```suggestion
        string='Invoice Alert', compute='_compute_display_invoice_alert'
```

File: addons/sale/wizard/sale_order_cancel.py [hunk 11]
```diff
@@ -1,20 +1,87 @@
...
+        raise UserError(_("Unable to post message, please configure the sender's email address."))
+
+    @api.model
+    def _default_author_id(self):
+        return self.env.user.partner_id
+
+    # origin
+    email_from = fields.Char(string='From', default=_default_email_from)
+    author_id = fields.Many2one(
+        'res.partner', string='Author', index=True, ondelete='set null', default=_default_author_id
```
Comment (Feyensv) [inline addons/sale/wizard/sale_order_cancel.py:26]: Sales-payment guideline: string shown to the user should be specified in double quotes :eyes:

File: addons/sale/wizard/sale_order_cancel.py [hunk 12]
```diff
@@ -1,20 +1,87 @@
...
-    display_invoice_alert = fields.Boolean('Invoice Alert', compute='_compute_display_invoice_alert')
+    display_invoice_alert = fields.Boolean(
+        'Invoice Alert', compute='_compute_display_invoice_alert'
+    )
+
+    @api.depends('order_id')
+    def _compute_recipient_ids(self):
+        for wizard in self:
+            wizard.recipient_ids += wizard.order_id.partner_id \
+                                 + wizard.order_id.message_partner_ids \
```
Comment (Feyensv) [inline addons/sale/wizard/sale_order_cancel.py:42]:
```suggestion
            wizard.recipient_ids = wizard.order_id.partner_id \
                                     | wizard.order_id.message_partner_ids \
```
union operator makes it clearer that there shouldn't be duplicates in the result :)

File: addons/sale/models/sale_order.py [hunk 13]
```diff
@@ -808,27 +808,42 @@ def action_unlock(self):
...
+                'name': _('Cancel %s', self.type_name),
+                'type': 'ir.actions.act_window',
                 'view_mode': 'form',
                 'res_model': 'sale.order.cancel',
-                'view_id': self.env.ref('sale.sale_order_cancel_view_form').id,
-                'type': 'ir.actions.act_window',
-                'context': {'default_order_id': self.id},
-                'target': 'new'
+                'target': 'new',
+                'context': ctx,
```
Comment (Feyensv) [inline addons/sale/models/sale_order.py:833]: keep the diff minimal here, keeping the view_id shouldn't change anything :)

File: addons/sale/wizard/sale_order_cancel.py [hunk 14]
```diff
@@ -1,20 +1,87 @@
 # -*- coding: utf-8 -*-
 # Part of Odoo. See LICENSE file for full copyright and licensing details.
 
-from odoo import api, fields, models
+from odoo import _, api, fields, models, tools
+from odoo.exceptions import UserError
```
Comment (Feyensv) [inline addons/sale/wizard/sale_order_cancel.py:5]:
```suggestion
from odoo import _, api, fields, models
from odoo.exceptions import UserError
from odoo.tools import formataddr
```

File: addons/sale/wizard/sale_order_cancel.py [hunk 15]
```diff
@@ -1,20 +1,87 @@
...
+                    engine='qweb',
+                )[wizard.order_id.id]
+
+    def action_send_mail_and_cancel(self):
+        self.ensure_one()
+        self.order_id.message_post(
+            subject=self.subject,
+            body=self.body,
+            message_type='comment',
+            email_from=self.author_id.email,
```
Comment (Feyensv) [inline addons/sale/wizard/sale_order_cancel.py:80]: why not self.email_from ?

**Review by Feyensv — COMMENTED** [review]
I think a `context` on an `action_cancel` call is still missing in `event_booth_sale` :
`order.sudo().action_cancel()`

File: addons/sale_purchase/models/sale_order.py [hunk 16]
```diff
@@ -23,8 +23,8 @@ def _action_confirm(self):
             order.order_line.sudo()._purchase_service_generation()
         return result
 
-    def action_cancel(self):
-        result = super(SaleOrder, self).action_cancel()
+    def _action_cancel(self):
+        result = super(SaleOrder, self)._action_cancel()
```
Comment (Feyensv) [inline addons/sale_purchase/models/sale_order.py:27]:
```suggestion
        result = super()._action_cancel()
```
let's clean this while we're on it :eyes:

File: addons/sale_loyalty/models/sale_order.py [hunk 17]
```diff
@@ -74,9 +74,9 @@ def action_confirm(self):
         self._send_reward_coupon_mail()
         return super(SaleOrder, self).action_confirm()
 
-    def action_cancel(self):
+    def _action_cancel(self):
         previously_confirmed = self.filtered(lambda s: s.state in ('sale', 'done'))
-        res = super(SaleOrder, self).action_cancel()
+        res = super(SaleOrder, self)._action_cancel()
```
Comment (Feyensv) [inline addons/sale_loyalty/models/sale_order.py:79]:
```suggestion
        res = super()._action_cancel()
```

File: addons/sale/models/sale_order.py [hunk 18]
```diff
@@ -806,29 +806,52 @@ def action_unlock(self):
...
+
+        For post-cancel operations, please only override `_action_cancel`.
+
+        note: self.ensure_one() if the wizard is shown.
+        """
         cancel_warning = self._show_cancel_wizard()
         if cancel_warning:
+            self.ensure_one()
+            template_id = self.env['ir.model.data']._xmlid_to_res_id(
+                'sale.mail_template_sale_cancellation', raise_if_not_found=False
```
Comment (Feyensv) [inline addons/sale/models/sale_order.py:819]: Does the following code even work if the template is deleted ?
(If yes, keep like that, if no, use a simple ref to make sure the record exists)

Comment (Demesmaeker) [reply addons/sale/models/sale_order.py:819]: Just tested it, it works :partying_face:

File: addons/sale_loyalty/models/sale_order.py [hunk 19]
```diff
@@ -74,9 +74,9 @@ def action_confirm(self):
         self._send_reward_coupon_mail()
         return super(SaleOrder, self).action_confirm()
 
-    def action_cancel(self):
+    def _action_cancel(self):
         previously_confirmed = self.filtered(lambda s: s.state in ('sale', 'done'))
-        res = super(SaleOrder, self).action_cancel()
+        res = super()._action_cancel()
```
Comment (Feyensv) [inline addons/sale_loyalty/models/sale_order.py:79]: @Williambraecky FYI, this should probably be fixed in 15.3 as well (Cf #81401, it seems you lost the change in your remastering of coupon ;))

File: addons/sale/wizard/sale_order_cancel_views.xml [hunk 20]
```diff
@@ -5,14 +5,46 @@
...
+                    <group col="2">
+                        <field name="subject" placeholder="Subject"/>
+                    </group>
+                    <field name="body"
+                           class="oe-bordered-editor"
+                           options="{'style-inline': true}"/>
+                </group>
                 <footer>
-                    <button string="Confirm" name="action_cancel" type="object" class="oe_highlight" data-hotkey="q"/>
-                    <button string="Cancel" class="btn btn-default" special="cancel" data-hotkey="z"/>
```
Comment (Feyensv) [inline addons/sale/wizard/sale_order_cancel_views.xml:15]: Why did you remove the hotkeys specification ?

File: addons/sale/wizard/sale_order_cancel_views.xml [hunk 21]
```diff
@@ -5,14 +5,46 @@
...
+                        <span id="display_invoice_alert"
+                              attrs="{'invisible': [('display_invoice_alert', '=', False)]}">
+                            Draft invoices for this order will be cancelled. <br/>
+                        </span>
+                    </div>
+                    <group col="2">
+                        <field name="recipient_ids"
+                               widget="many2many_tags_email"
+                               context="{'force_email':True,
+                                         'show_email':True,
```
Comment (Feyensv) [inline addons/sale/wizard/sale_order_cancel_views.xml:26]:
```suggestion
                               context="{'force_email': True,
                                         'show_email': True,
```
:D

File: addons/sale_purchase/models/sale_order.py [hunk 22]
```diff
@@ -23,8 +23,8 @@ def _action_confirm(self):
             order.order_line.sudo()._purchase_service_generation()
         return result
 
-    def action_cancel(self):
-        result = super(SaleOrder, self).action_cancel()
+    def _action_cancel(self):
+        result = super()._action_cancel()
```
Comment (Feyensv) [inline addons/sale_purchase/models/sale_order.py:27]: Please make a specific commit for this `FIX` (so that I can backport it in 13+) :).

**Review by Feyensv — COMMENTED** [review]
Really should be more attentive in my reviews (and stop making n mini reviews :pensive:)

File: addons/sale_loyalty/models/sale_order.py (hunk 19 above)
Comment (Williambraecky) [reply addons/sale_loyalty/models/sale_order.py:79]: Here: https://github.com/odoo/odoo/pull/90861 :)

File: addons/sale/wizard/sale_order_cancel_views.xml (hunk 20 above)
Comment (Demesmaeker) [reply addons/sale/wizard/sale_order_cancel_views.xml:15]: 2 buttons => 3, it was unclear which shortcut should be on which button, we let the auto-assigned buttons instead.

**Review by Feyensv — APPROVED** [review]
@robodoo r+ rebase-ff
